cloudant-sync-datastore-android/src/main/java/com/cloudant/sync/replication/ReplicationService.java

```diff
@@ -0,0 +1,323 @@
+        /** Callback to indicate that an operation has completed. This should only be used in
+         * automated tests. */
+        void operationComplete();
+    }
+
+    public interface ReplicationCompleteListener {
+        /**
+         * Callback to indicate that all replications specified by {@link #getReplicators(Context)}
+         * are complete.
+         */
+        void allReplicationsComplete();
+
+        /** Callback to indicate that the individual replication with the given {@code id} is
+         * complete.
+         * @param id the {@code id} number associated wit the replication that has completed. See
+         *           {@link ReplicatorBuilder#withId(int)}.
+         */
+        void replicationComplete(int id);
+
+        /** Callback to indicate that the individual replication with the given {@code id} has
+         * errored.
+         * @param id the {@code id} number associated wit the replication that has completed. See
+         *           {@link ReplicatorBuilder#withId(int)}.
+         */
+        void replicationErrored(int id);
+    }
+
+    /**
+     * A simple {@link com.cloudant.sync.replication.ReplicationService.ReplicationCompleteListener}
+     * to save clients having to override every method if they are only interested in a subset of
+     * the events.
+     */
+    public static class SimpleReplicationCompleteListener implements ReplicationCompleteListener {
+        @Override
+        public void allReplicationsComplete() {
+        }
+
+        @Override
+        public void replicationComplete(int id) {
+        }
+
+        @Override
+        public void replicationErrored(int id) {
+        }
+    }
+
+    // A binder to allow components running in the same process to bind to this Service.
+    public class LocalBinder extends Binder {
+        public ReplicationService getService() {
+            return ReplicationService.this;
+        }
+    }
+
+    // Handler that receives messages from the thread.
+    protected class ServiceHandler extends Handler {
+        public ServiceHandler(Looper looper) {
+            super(looper);
+        }
+
+        @Override
+        public void handleMessage(Message msg) {
+            // Process the commands passed in msg.arg2.
+            switch (msg.arg2) {
+                case COMMAND_START_REPLICATION:
+                    startReplications();
+                    notifyTestOperationComplete();
+                    break;
+                case COMMAND_STOP_REPLICATION:
+                    stopReplications();
+                    notifyTestOperationComplete();
+                    break;
+            }
+
+            // Get the Intent used to start the service and release the WakeLock if there is one.
+            // Calling completeWakefulIntent is safe even if there is no wakelock held.
+            Intent intent = msg.getData().getParcelable(EXTRA_INTENT);
+            WakefulBroadcastReceiver.completeWakefulIntent(intent);
```

Comment:
I wonder if this should be in a `finally` with the other parts of the method in a `try` to ensure that the `WakeLock` is released even if there is some kind of `RuntimeException` from any of the starting and stopping code.